File: 14.longest-common-prefix/cpp/tests/longest_common_prefix_test.cc
#include <gtest/gtest.h>
#include "../src/longest_common_prefix.h"

void processTestCase(vector<string> input, string expected) {
	Solution* solution_ptr = new Solution();

	// since `solution_ptr` is a pointer which points to Solution class
	// just use `->` to access the class member
	string actual = solution_ptr->longestCommonPrefix(input);
	ASSERT_EQ(actual, expected);
}

TEST(LongestCommonPrefixTest, HaveLongestCommonPrefix) {
	vector<string> input = { "flower", "flow", "flight" };
	string expected = "fl";
	processTestCase(input, expected);
}

TEST(LongestCommonPrefixTest, NoLongestCommonPrefix) {
	vector<string> input = { "dog", "racecar", "car" };
	string expected = "";
	processTestCase(input, expected);
}

File: 14.longest-common-prefix/rust/tests/longest_common_prefix_test.rs
use leetcode::Solution;

#[test]
fn have_longest_common_prefix() {
	let input: Vec<String> = vec![ "flower", "flow", "flight" ]
							.into_iter()
							.map(String::from)
							.collect();
	let expected: &str = "fl";

	process_test_case(&input, expected);
}

#[test]
fn no_longest_common_prefix() {
	let input: Vec<String> = vec![ "dog", "racecar", "car" ]
							.into_iter()
							.map(String::from)
							.collect();
	let expected: &str = "";

	process_test_case(&input, expected);
}

fn process_test_case(input: &Vec<String>, expected: &str) {
	let solution: Solution = Solution::new();
	let actual: String = solution.longest_common_prefix(input);
	assert_eq!(&actual, expected);
}
